exercises/pick_place/web-template/gui.py

```diff
@@ -0,0 +1,247 @@
+
+        return payload
+
+    # Function to prepare image payload
+    # Encodes the image as a JSON string and sends through the WS
+    def payloadLeftImage(self):
+        self.left_image_show_lock.acquire()
+        left_image_to_be_shown_updated = self.left_image_to_be_shown_updated
+        left_image_to_be_shown = self.left_image_to_be_shown
+        self.left_image_show_lock.release()
+
+        image = left_image_to_be_shown
+        payload = {'image': '', 'shape': ''}
+
+        if not left_image_to_be_shown_updated:
+            return payload
+
+        shape = image.shape
+        frame = cv2.imencode('.JPEG', image)[1]
+        encoded_image = base64.b64encode(frame)
+
+        payload['image'] = encoded_image.decode('utf-8')
+        payload['shape'] = shape
+
+        self.left_image_show_lock.acquire()
+        self.left_image_to_be_shown_updated = False
+        self.left_image_show_lock.release()
+
+        return payload
+
+    # Function for student to call
+    def showImage(self, image):
+        self.image_show_lock.acquire()
+        self.image_to_be_shown = image
+        self.image_to_be_shown_updated = True
+        self.image_show_lock.release()
+
+    # Function for student to call
+    def showLeftImage(self, image):
+        self.left_image_show_lock.acquire()
+        self.left_image_to_be_shown = image
+        self.left_image_to_be_shown_updated = True
+        self.left_image_show_lock.release()
+
+    # Function to get the client
+    # Called when a new client is received
+    def get_client(self, client, server):
+        self.client = client
+        
+    # Function to get value of Acknowledge
+    def get_acknowledge(self):
+        self.acknowledge_lock.acquire()
+        acknowledge = self.acknowledge
+        self.acknowledge_lock.release()
+        
+        return acknowledge
+        
+    # Function to get value of Acknowledge
+    def set_acknowledge(self, value):
+        self.acknowledge_lock.acquire()
+        self.acknowledge = value
+        self.acknowledge_lock.release()
+        
+    # Update the gui
+    def update_gui(self):
+        # Payload Image Message
+        payload = self.payloadImage()
+        self.payload["image"] = json.dumps(payload)
+        
+        message = "#gui" + json.dumps(self.payload)
+        self.server.send_message(self.client, message)
+
+        # Payload Left Image Message
+        left_payload = self.payloadLeftImage()
+        self.left_payload["image"] = json.dumps(left_payload)
+
+        message = "#gul" + json.dumps(self.left_payload)
+        self.server.send_message(self.client, message)
+            
+    # Function to read the message from websocket
+    # Gets called when there is an incoming message from the client
+    def get_message(self, client, server, message):
+        # Acknowledge Message for GUI Thread
+        if message[:4] == "#ack":
+            self.set_acknowledge(True)
+        elif message[:4] == "#tur":
```

Comment:
If we are referring turtlebot template to make the pick and place exercise work, please make sure to update variable names / code specific to pick_place. 

There is a lot of turtle bot exercise specific redundant code which came for pick_place exercise as well. Please find those specific pieces and update the exercise.